### ISSUE TITLE
Make tests deterministics

### ISSUE THREAD

**Issue by schreter** [datafuselabs/openraft#206]
This depends on https://github.com/datafuselabs/openraft/issues/204 and https://github.com/datafuselabs/openraft/issues/205.

Currently, tests use real-time timers, which quite often cause non-deterministic results and as a consequence test failures (false positives).

Instead, create a test `Runtime`, which abstracts the "real" runtime and timing and provides a top-level simple executor "ticking" the simulated time. I.e., it will run the main `Future` of the test to the completion by polling it repeatedly and incrementing simulated time between polls (which indicate a blocking situation) by some sensible amount, such as 10ms.

Timeouts would be driven by the test `Runtime` timer instead of real-time. This way, it's possible to simulate all kinds of situations easily and deterministically prove that the algorithms properly resolve various error situations.

Aside from that, since tests won't be running in real-time, I estimate they will be two orders of magnitude faster (i.e., tens of milliseconds instead of seconds).

**Comment by drmingdrmer** [comment]
I just changed this post a bit to make it easy to read:D

**Comment by drmingdrmer** [comment]
> Instead, create a test `Runtime`, which abstracts the "real" runtime and timing and provides a top-level simple executor "ticking" the simulated time. I.e., it will run the main `Future` of the test to the completion by polling it repeatedly and incrementing simulated time between polls (which indicate a blocking situation) by some sensible amount, such as 10ms.

Great Idea!!!
In short, replace tokio runtime with a customized one, right?

> Timeouts would be driven by the test `Runtime` timer instead of real-time. This way, it's possible to simulate all kinds of situations easily and deterministically prove that the algorithms properly resolve various error situations.

One of the timeout values is generated randomly:
https://github.com/datafuselabs/openraft/blob/46644c8409c7dff627c991cc65849507ad5265b3/openraft/src/core/mod.rs#L425-L435
I guess in the design with the **Runtime for test**, this has to be replaced with a method of the runtime, right?
Such as `runtiiie.rand_sleep(Duration::from_millis(10))`, to make it deterministic.


> Aside from that, since tests won't be running in real-time, I estimate they will be two orders of magnitude faster (i.e., tens of milliseconds instead of seconds).

:DDD

**Comment by schreter** [comment]
> In short, replace tokio runtime with a customized one, right?

Yes. There are just a couple primitives that are needed in the implementation. Implementing them trivially for test purposes will allow to make tests deterministic.

> One of the timeout values is generated randomly
> I guess in the design with the Runtime for test, this has to be replaced with a method of the runtime, right?

I know. Yes, that one can be also abstracted through the runtime and made deterministic for tests.

> I just changed this post a bit to make it easy to read:D

Hehe :-). I'm not a GitHub expert, so whatever fits the best.
Surely there is a guide how to make such nice links there, but I didn't actively search for it yet.

**Comment by drmingdrmer** [comment]
> Hehe :-). I'm not a GitHub expert, so whatever fits the best.
> Surely there is a guide how to make such nice links there, but I didn't actively search for it yet.

I did not read the doc, but one day I found github turns `- blabla #<issue_number>` into a link with the text of the issue title, which is handy to use.:DDD

**Comment by pchrysostomou** [comment]
@drmingdrmer is #206 still wanted now that tests-turmoil covers seeded cluster fuzzing? If so, is the target running `tests/tests` on virtual time, or something else?

To find out, I built a local prototype: a standalone `rt-sim` crate with a virtual clock that jumps to the next timer, FIFO scheduling and its own `Watch`. It passes 42 of the 45 `Suite` tests and all 5 `DetsimSuite` tests in 0.00s, against about 0.4s for tokio, monoio and compio. Its scheduling trace was byte-identical across 7 processes, and a deliberately LIFO run queue produced a 484-line diff.

The three failures are `test_sleep`, `test_watch_changed_marks_as_seen` and `test_watch_multiple_borrow_then_changed`, which measure elapsed time with `std::time::Instant` in `rt/src/testing.rs`. Switching them to `Rt::Instant` is a three-line change: tokio, monoio and compio still pass, and rt-sim then passes `Suite::test_all` unmodified. Would that be welcome as a small first PR?

Worth knowing either way: with tokio's `rt` feature on, which openraft enables by default, `tokio::sync::watch` puts each waiter in one of eight slots chosen by tokio's RNG, so eight waiters woke in 12 different orders across 12 runs. No other tokio sync primitive touches the RNG.

Open questions: a standalone crate beside rt-monoio, or a feature in `rt`? Jump to the next timer, or tick by 10ms as the issue suggests? And must it support `single-threaded`? Wakers need `Send + Sync`, which means keeping task futures out of shared state.

**Comment by drmingdrmer** [comment]
Thanks. #206 is still useful for running targeted `tests/tests` cases with deterministic virtual time; `tests-turmoil` covers a different fuzzing role.

Please open the runtime-suite clock cleanup first. Also keep `test_sleep` distinct from `test_instant`, for example by checking that a non-zero sleep is initially pending.

Please publish `rt-sim` as a draft PR before we choose the architecture. The draft should run a representative integration test and demonstrate repeatable full OpenRaft traces. `AsyncRuntime` does not control the shuffle RNG used by `futures_util::select!`, which our Turmoil harness currently reseeds.

If we proceed with `rt-sim`, I prefer a standalone crate and jumping to the next timer after all runnable tasks have drained. `single-threaded` support can follow the first working integration.
